cpp/src/io/parquet/writer_impl.hpp

```diff
@@ -68,57 +66,71 @@ class writer::impl {
+                rmm::mr::device_memory_resource* mr,
+                rmm::cuda_stream_view stream);
+
+  /**
+   * @brief Destructor to complete any incomplete write and release resources.
+   */
+  ~impl();
+
+  /**
+   * @brief Initializes the states before writing.
+   */
+  void init_state();
 
   /**
    * @brief Write an entire dataset to parquet format.
    *
    * @param table The set of columns
-   * @param metadata The metadata associated with the table
    * @param return_filemetadata If true, return the raw parquet file metadata
    * @param column_chunks_file_path Column chunks file path to be set in the raw output metadata
-   * @param stream CUDA stream used for device memory operations and kernel launches.
+   *
    * @return unique_ptr to FileMetadata thrift message if requested
    */
   std::unique_ptr<std::vector<uint8_t>> write(table_view const& table,
-                                              const table_metadata* metadata,
                                               bool return_filemetadata,
-                                              const std::string& column_chunks_file_path,
-                                              std::vector<uint8_t> const& decimal_precisions,
-                                              rmm::cuda_stream_view stream);
-
-  /**
-   * @brief Begins the chunked/streamed write process.
-   *
-   * @param[in] pq_chunked_state Internal state maintained between chunks.
-   */
-  void write_chunked_begin(pq_chunked_state& state);
+                                              const std::string& column_chunks_file_path);
 
   /**
-   * @brief Writes a single subtable as part of a larger parquet file/table write.
+   * @brief Writes a single subtable as part of a larger parquet file/table write,
+   * normally used for chunked writing.
    *
    * @param[in] table The table information to be written
-   * @param[in] pq_chunked_state Internal state maintained between chunks.
-   * boundaries.
    */
-  void write_chunk(table_view const& table, pq_chunked_state& state);
+  void write(table_view const& table);
 
   /**
    * @brief Finishes the chunked/streamed write process.
    *
-   * @param[in] pq_chunked_state Internal state maintained between chunks.
-   * @param return_filemetadata If true, return the raw parquet file metadata
-   * @param column_chunks_file_path Column chunks file path to be set in the raw output metadata
+   * @param[in] return_filemetadata If true, return the raw parquet file metadata
+   * @param[in] column_chunks_file_path Column chunks file path to be set in the raw output metadata
    * @return unique_ptr to FileMetadata thrift message if requested
    */
-  std::unique_ptr<std::vector<uint8_t>> write_chunked_end(
-    pq_chunked_state& state,
-    bool return_filemetadata                   = false,
-    const std::string& column_chunks_file_path = "");
+  std::unique_ptr<std::vector<uint8_t>> close(bool return_filemetadata                   = false,
```

Comment:
Maybe it's a good opportunity to replace this bool parameter, since we're avoiding them in the public API.
I'm also trying to figure out if we can combine the two params, but I'm not 100% clear on the semantics.

Reply:
As per my observation if column_chunks_file_path is not empty then only is true, so we can just keep column_chunks_file_path.